react/components/RenderProvider.tsx

```diff
@@ -622,45 +633,18 @@ class RenderProvider extends Component<Props, RenderProviderState> {
 
   public prefetchDefaultPages = async (routeIds: string[]) => {
     const {
-      runtime,
-      runtime: { renderMajor },
+      runtime: { query },
     } = this.props
-    const {
-      culture: { locale },
-      pages,
-    } = this.state
 
-    const {
-      components: defaultComponents,
-      extensions: defaultExtensions,
-      messages: defaultMessages,
-    } = await fetchDefaultPages({
-      apolloClient: this.apolloClient,
-      locale,
-      pages,
-      renderMajor,
-      routeIds,
-    })
-
-    await Promise.all(
-      Object.keys(defaultComponents).map((component: string) => {
-        const componentsAssetsMap = traverseComponent(
-          defaultComponents,
-          component
-        )
-        return prefetchAssets(runtime, componentsAssetsMap)
-      })
-    )
-
-    this.setState(({ components, messages }) => ({
-      components: {
-        ...defaultComponents,
-        ...this.state.components,
-        ...components,
-      },
-      defaultExtensions,
-      messages: { ...defaultMessages, ...this.state.messages, ...messages },
-    }))
+    const { prefetchRoutes } = this.state
+    const disablePrefetch =
+      query &&
+      DISABLE_PREFETCH_PAGES in query &&
+      query[DISABLE_PREFETCH_PAGES] !== 'false'
+    if (!disablePrefetch) {
+      routeIds.forEach(routeId => prefetchRoutes.add(routeId))
+      this.setState({ prefetchRoutes })
```

Comment:
I think we should never use `setState` here. Prefetching should be lazy.

`prefetchRoutes` can be just a private field of `RenderProvider`

One concern is that `prefetchDefaultPages` is a function always available in runtime context, but it only works if you call it before RenderProvider is mounted. If that is the behavior we want we should at least log a warning and do nothing if someone calls it after `componentDidMount`.

Reply:
Good point! I think we should only execute prefetchPages once, it doesn't make sense to call it after RenderProvider is already rendered.
I'm throwing an exception when someone calls it after render provider have been already rendered.